data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/AvroConsumer.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.kafka.consumer;
+
+import org.apache.avro.generic.GenericRecord;
+
+import org.apache.kafka.clients.consumer.CommitFailedException;
+import org.apache.kafka.clients.consumer.ConsumerRebalanceListener;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.opensearch.dataprepper.metrics.PluginMetrics;
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.plugins.kafka.configuration.KafkaSourceConfig;
+import org.opensearch.dataprepper.plugins.kafka.configuration.TopicsConfig;
+import org.opensearch.dataprepper.plugins.kafka.source.KafkaSourceBufferAccumulator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Collection;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+
+/**
+ * * A helper class which helps to process the avro records and write them into the buffer.
+ * Offset handling and consumer re-balance are also being handled
+ */
+
+@SuppressWarnings("deprecation")
+public class AvroConsumer implements KafkaSourceSchemaConsumer<String, GenericRecord>, ConsumerRebalanceListener {
+    private static final Logger LOG = LoggerFactory.getLogger(AvroConsumer.class);
+    private Map<TopicPartition, OffsetAndMetadata> offsetsToCommit = new HashMap<>();
+    private long lastReadOffset = 0L;
+    private KafkaConsumer<String, GenericRecord> kafkaAvroConsumer;
+    private volatile long lastCommitTime = System.currentTimeMillis();
+
+    @Override
+    public void consumeRecords(final KafkaConsumer<String, GenericRecord> consumer, final AtomicBoolean status,
+                               final Buffer<Record<Object>> buffer, final TopicsConfig topicConfig,
+                               final KafkaSourceConfig kafkaSourceConfig, PluginMetrics pluginMetrics,
+                               final String schemaType) {
+        KafkaSourceBufferAccumulator kafkaSourceBufferAccumulator = new KafkaSourceBufferAccumulator(topicConfig,kafkaSourceConfig,
+                schemaType,pluginMetrics);
+        kafkaAvroConsumer = consumer;
+        try {
+            consumer.subscribe(Arrays.asList(topicConfig.getName()));
+            while (!status.get()) {
```

Comment:
There is quite a bit of nesting in this loop, please create separate methods for better readability and it will also be easier to test.

Reply:
Resolved